presto-tests/src/test/java/com/facebook/presto/tests/TestQueryManager.java

```diff
@@ -56,6 +60,30 @@ public void tearDown()
         queryRunner = null;
     }
 
+    @Test(timeOut = 100_000L)
+    public void testFailedQueryCleanUp()
+            throws Exception
+    {
+        QueryManager queryManager = queryRunner.getCoordinator().getQueryManager();
+        assertTrue(queryManager instanceof SqlQueryManager);
+        SqlQueryManager sqlQueryManager = (SqlQueryManager) queryManager;
+        // no active transaction before query starts
+        assertTrue(sqlQueryManager.getTransactionManager().getAllTransactionInfos().isEmpty());
+        QueryId queryId = sqlQueryManager.createQueryId();
+        queryManager.createQuery(
+                queryId,
+                new TestingSessionContext(TEST_SESSION),
+                "SELECT non_exist_column FROM lineitem")
+                .get();
+
+        QueryInfo queryInfo = sqlQueryManager.getFullQueryInfo(queryId);
+        assertEquals(queryInfo.getState(), FAILED);
+        assertNotNull(queryInfo.getFailureInfo());
+        TimeUnit.MILLISECONDS.sleep(50);
```

Comment:
Why is there a sleep call here? Does the correctness of the test depend on this call?

Reply:
This test will pass without sleep on my laptop. However, clean up of metadata is async job, it is better to wait for a bit.

Reply:
Can we get a future on that metadata cleanup asyn job and wait until it's done? Sleeping is fragile, this test can fail arbitrarily depending on the interleaving of the async job and the thread running this test.

Reply:
@nezihyigitbasi After further investigation with @arhimondr, we confirmed sleep() is not necessary since cleanup is synchronous. I have removed it from test.